File: src/Java/1. First Course/5. Fourth Hole/src/main/refactoringgolf/store/Manager.java
package refactoringgolf.store;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class Manager extends Employee {

	protected Set<Employee> subordinates = new HashSet<Employee>();

	public Manager(String firstName, String lastName, float fixedSalary) {
		super(firstName, lastName, fixedSalary);
	}

	public Set<Employee> getSubordinates() {
		return Collections.unmodifiableSet(subordinates);
	}

	public void addSubordinate(Employee subordinate) {
		subordinates.add(subordinate);
		subordinate.manager = this;
	}

	public void removeSubordinate(Employee subordinate) {
		subordinates.remove(subordinate);
		subordinate.manager = null;
	}
	
	@Override
	protected float salaryBenefits() {
		return this.subordinates.size() * 20;
	}
}

File: src/Net/1. First Course/4. Fourth Hole/RefactoringGolf.Store/Manager.cs
namespace RefactoringGolf.Store
{
    using System.Collections.Generic;
    using System.Linq;

    public class Manager : Employee
    {
        protected IList<Employee> subordinates = new List<Employee>();

        public Manager(string firstName, string lastName, decimal fixedSalary)
            : base(firstName, lastName, fixedSalary)
        {
        }

        protected override decimal SalaryBenefits()
        {
            return this.subordinates.Count * 20;
        }

        public IEnumerable<Employee> Subordinates
        {
            get { return subordinates.ToArray(); }
        }

        public void AddSubordinate(Employee subordinate)
        {
            subordinates.Add(subordinate);
            subordinate.Manager = this;
        }

        public void RemoveSubordinate(Employee subordinate)
        {
            subordinates.Remove(subordinate);
            subordinate.Manager = null;
        }
    }
}